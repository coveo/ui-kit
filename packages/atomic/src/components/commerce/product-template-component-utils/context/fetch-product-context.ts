import type {Product} from '@coveo/headless/commerce';
import {fetchItemContext} from '@/src/components/common/item-list/fetch-item-context';

/**
 * Retrieves `Product` on a rendered `atomic-product`.
 *
 * This utility function is useful for building custom product template elements, see [Create a Product List](https://docs.coveo.com/en/atomic/latest/cc-search/create-custom-components/native-components/#custom-product-template-component-example) for more information.
 *
 * You should use the utility function in the [`connectedCallback` lifecycle method](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements#using_the_lifecycle_callbacks).
 *
 * @param element - The element that the event is dispatched to, which must be the child of a rendered `atomic-product`.
 * @returns A promise that resolves on initialization of the parent `atomic-product` element, or rejects when there is no parent `atomic-product` element.
 */
export function fetchProductContext<T extends Product>(element: Element) {
  return fetchItemContext<T>(element, 'atomic-product');
}
