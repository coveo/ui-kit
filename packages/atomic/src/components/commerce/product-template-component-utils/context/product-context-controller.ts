import type {InteractiveProduct, Product} from '@coveo/headless/commerce';
import type {LitElement} from 'lit';
import type {InteractiveItemContextEvent} from '@/src/components/common/item-list/context/interactive-item-context-controller';
import {
  ItemContextController,
  type ItemContextEvent,
} from '@/src/components/common/item-list/context/item-context-controller';

/**
 * Creates a [Lit reactive controller](https://lit.dev/docs/composition/controllers/) for managing product context in product template components.
 *
 * @param host - The Lit component instance
 * @param options - Configuration options
 * @returns ItemContextController instance configured for atomic-product
 *
 * @example
 * ```typescript
 * @customElement('my-product-component')
 * export class MyProductComponent extends LitElement {
 *   private productController = createProductContextController(this);
 *
 *   @state() product!: Product;
 *
 *   render() {
 *     this.product = this.productController.item;
 *     return html`<div>${this.product?.ec_name}</div>`;
 *   }
 * }
 * ```
 */
export function createProductContextController(
  host: LitElement & {error: Error | null},
  options: {folded?: boolean} = {}
): ItemContextController<Product> {
  return new ItemContextController<Product>(host, {
    parentName: 'atomic-product',
    folded: options.folded ?? false,
  });
}

export type ProductContextEvent<T = Product> = ItemContextEvent<T>;
export type InteractiveProductContextEvent<
  T extends InteractiveProduct = InteractiveProduct,
> = InteractiveItemContextEvent<T>;
