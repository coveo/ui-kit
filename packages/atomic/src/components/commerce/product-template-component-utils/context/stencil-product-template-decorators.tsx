import {Product} from '@coveo/headless/commerce';
import {itemContext} from '@/src/components/common/item-list/stencil-item-decorators';

/**
 * Retrieves `Product` on a rendered `atomic-product`.
 *
 * This method is useful for building custom product template elements, see [Create a Product List](https://docs.coveo.com/en/atomic/latest/cc-search/create-custom-components/native-components/#custom-product-template-component-example) for more information.
 *
 * You should use the method in the [connectedCallback lifecycle method](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements#using_the_lifecycle_callbacks).
 *
 * @param element - The element that the event is dispatched to, which must be the child of a rendered "atomic-product".
 * @returns A promise that resolves on initialization of the parent "atomic-product" element, or rejects when there is no parent "atomic-product" element.
 *
 * @deprecated should only be used for Stencil components. For Lit components, use `fetchProductContext` from \@/src/components/commerce/product-template-component-utils/context/fetch-product-context.ts
 */
export function productContext<T extends Product>(element: Element) {
  return itemContext<T>(element, 'atomic-product');
}
