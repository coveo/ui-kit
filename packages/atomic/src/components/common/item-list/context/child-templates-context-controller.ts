import {ItemTemplateProvider} from '@/src/components/common/item-list/item-template-provider';
import {LitElementWithError} from '@/src/decorators/types';
import {buildCustomEvent} from '@/src/utils/event-utils';
import {ReactiveController, ReactiveControllerHost} from 'lit';

const childTemplatesContextEventName = 'atomic/resolveChildTemplates';

/**
 * A reactive controller that manages child template context data from parent components.
 * Handles fetching ItemTemplateProvider via custom events and manages the template provider state.
 */
export class ChildTemplatesContextController implements ReactiveController {
  private host: ReactiveControllerHost & LitElementWithError;
  private _itemTemplateProvider: ItemTemplateProvider | null = null;

  constructor(host: ReactiveControllerHost & LitElementWithError) {
    this.host = host;
    host.addController(this);
  }

  get itemTemplateProvider(): ItemTemplateProvider | null {
    return this._itemTemplateProvider;
  }

  hostConnected(): void {
    this._resolveChildTemplatesContext();
  }

  hostUpdated(): void {
    if (!this._itemTemplateProvider) {
      this._resolveChildTemplatesContext();
    }
  }

  private _resolveChildTemplatesContext(): void {
    const event = buildCustomEvent(
      childTemplatesContextEventName,
      (itemTemplateProvider?: ItemTemplateProvider) => {
        if (this._itemTemplateProvider) {
          return;
        }

        this._itemTemplateProvider = itemTemplateProvider || null;
        this.host.requestUpdate();
      }
    );

    const canceled = this.host.dispatchEvent(event);
    if (canceled) {
      this._itemTemplateProvider = null;
      this.host.requestUpdate();
    }
  }
}

type ChildTemplatesContextEventHandler = (
  itemTemplateProvider?: ItemTemplateProvider
) => void;
export type ChildTemplatesContextEvent =
  CustomEvent<ChildTemplatesContextEventHandler>;
