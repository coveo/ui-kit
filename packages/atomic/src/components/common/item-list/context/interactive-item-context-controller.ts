import {AnyItem} from '@/src/components/common/interface/item.js';
import {LitElementWithError} from '@/src/decorators/types';
import {buildCustomEvent} from '@/src/utils/event-utils';
import {ReactiveController, ReactiveControllerHost} from 'lit';

const interactiveItemContextEventName = 'atomic/resolveInteractiveResult';

/**
 * A reactive controller that manages interactive item context data from parent components.
 * Handles fetching interactive item data via custom events.
 */
export class InteractiveItemContextController<T extends AnyItem = AnyItem>
  implements ReactiveController
{
  private host: ReactiveControllerHost & LitElementWithError;
  private _interactiveItem: T | null = null;

  constructor(host: ReactiveControllerHost & LitElementWithError) {
    this.host = host;
    host.addController(this);
  }

  get interactiveItem(): T | null {
    return this._interactiveItem;
  }

  hostConnected(): void {
    this._resolveInteractiveItemContext();
  }

  private _resolveInteractiveItemContext(): void {
    const event = buildCustomEvent(
      interactiveItemContextEventName,
      (item: AnyItem) => {
        this._interactiveItem = item as T;
        this.host.requestUpdate();
      }
    );

    this.host.dispatchEvent(event);
  }
}

export type InteractiveItemContextEvent<T extends AnyItem = AnyItem> =
  CustomEvent<(interactiveItem: T) => void>;
