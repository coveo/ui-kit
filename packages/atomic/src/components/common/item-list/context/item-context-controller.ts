import type {LitElement, ReactiveController, ReactiveControllerHost} from 'lit';
import type {InitializableComponent} from '@/src/decorators/types.js';
import {buildCustomEvent} from '@/src/utils/event-utils';

type LitElementWithError = Omit<
  Pick<InitializableComponent, 'error'>,
  'error'
> &
  LitElement & {
    error: Error | null;
  };

const itemContextEventName = 'atomic/resolveResult';

export class MissingParentError extends Error {
  constructor(elementName: string, parentName: string) {
    super(
      `The "${elementName}" element must be the child of an "${parentName}" element.`
    );
  }
}

function extractFolded<T = Record<string, unknown>>(
  item: Record<string, unknown>,
  returnFolded: boolean
): T {
  if (returnFolded) {
    if ('children' in item) {
      return item as T;
    } else {
      return {children: [], result: item} as T;
    }
  }

  if ('children' in item && 'result' in item) {
    return item.result as T;
  }
  return item as T;
}

/**
 * A reactive controller that manages item context data from parent components.
 * Handles fetching item data via custom events and manages error states.
 */
export class ItemContextController<T = Record<string, unknown>>
  implements ReactiveController
{
  private host: ReactiveControllerHost & LitElementWithError;
  private parentName: string;
  private folded: boolean;
  private _item: T | null = null;
  private _error: MissingParentError | null = null;

  constructor(
    host: ReactiveControllerHost & LitElementWithError,
    options: {parentName?: string; folded?: boolean} = {}
  ) {
    this.host = host;
    this.parentName = options.parentName ?? 'atomic-result';
    this.folded = options.folded ?? false;
    host.addController(this);
  }

  get item(): T | null {
    return this._error ? null : this._item;
  }

  get error(): MissingParentError | null {
    return this._error;
  }

  get hasError(): boolean {
    return this._error !== null;
  }

  hostConnected(): void {
    this._resolveItemContext();
  }

  private _resolveItemContext(): void {
    const event = buildCustomEvent(
      itemContextEventName,
      (item: Record<string, unknown>) => {
        this._item = extractFolded<T>(item, this.folded);
        this._error = null;
        this.host.error = null;
        this.host.requestUpdate();
      }
    );

    const canceled = this.host.dispatchEvent(event);
    if (canceled) {
      const elementName = (this.host as Element).nodeName.toLowerCase();
      this._error = new MissingParentError(elementName, this.parentName);
      this._item = null;
      this.host.error = this._error;
      this.host.requestUpdate();
    }
  }
}

type ItemContextEventHandler<T> = (item: T) => void;
export type ItemContextEvent<T> = CustomEvent<ItemContextEventHandler<T>>;
