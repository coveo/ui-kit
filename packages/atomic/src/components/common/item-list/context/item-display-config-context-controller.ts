import type {LitElement, ReactiveController, ReactiveControllerHost} from 'lit';
import type {
  ItemDisplayDensity,
  ItemDisplayImageSize,
} from '@/src/components/common/layout/display-options';
import {buildCustomEvent} from '@/src/utils/event-utils';

const itemDisplayConfigContextEventName = 'atomic/resolveResultDisplayConfig';

export type DisplayConfig = {
  density: ItemDisplayDensity;
  imageSize: ItemDisplayImageSize;
};

/**
 * A reactive controller that manages display configuration context from parent components.
 * Handles fetching display config data via custom events.
 */
export class ItemDisplayConfigContextController implements ReactiveController {
  private host: ReactiveControllerHost & LitElement;
  private _config: DisplayConfig | null = null;

  constructor(host: ReactiveControllerHost & LitElement) {
    this.host = host;
    host.addController(this);
  }

  get config(): DisplayConfig | null {
    return this._config;
  }

  hostConnected(): void {
    this._resolveDisplayConfig();
  }

  hostUpdated(): void {
    this._resolveDisplayConfig();
  }

  private _resolveDisplayConfig(): void {
    const event = buildCustomEvent(
      itemDisplayConfigContextEventName,
      (config: DisplayConfig) => {
        this._config = config;
        this.host.requestUpdate();
      }
    );

    const canceled = this.host.dispatchEvent(event);
    if (canceled) {
      this._config = null;
    }
  }
}

type ItemDisplayConfigContextEventHandler = (config: DisplayConfig) => void;
export type ItemDisplayConfigContextEvent =
  CustomEvent<ItemDisplayConfigContextEventHandler>;
