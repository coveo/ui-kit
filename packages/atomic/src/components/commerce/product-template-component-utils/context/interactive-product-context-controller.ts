import type {InteractiveProduct} from '@coveo/headless/commerce';
import type {LitElement} from 'lit';
import {InteractiveItemContextController} from '@/src/components/common/item-list/context/interactive-item-context-controller';

/**
 * Creates a [Lit reactive controller](https://lit.dev/docs/composition/controllers/) for managing interactive product context in product template components.
 *
 * @param host - The Lit component instance
 * @returns InteractiveItemContextController instance
 *
 * @example
 * ```typescript
 * @customElement('my-interactive-product-component')
 * export class MyInteractiveProductComponent extends LitElement {
 *   private interactiveProductController = createInteractiveProductContextController(this);
 *
 *   @state() product!: Product;
 *
 *   render() {
 *     this.product = this.interactiveProductController.interactiveItem;
 *     return html`<div>${this.product?.ec_name}</div>`;
 *   }
 * }
 * ```
 */
export function createInteractiveProductContextController(
  host: LitElement & {error: Error}
): InteractiveItemContextController<InteractiveProduct> {
  return new InteractiveItemContextController<InteractiveProduct>(host);
}
