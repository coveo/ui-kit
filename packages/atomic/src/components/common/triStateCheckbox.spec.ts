import {renderFunctionFixture} from '@/vitest-utils/testing-helpers/fixture';
import {page} from '@vitest/browser/context';
import '@vitest/browser/matchers.d.ts';
import {html} from 'lit';
import {vi, expect, describe, it} from 'vitest';
import {
  renderTriStateCheckbox,
  TriStateCheckboxProps,
} from './triStateCheckbox';

describe('renderTriStateCheckbox', () => {
  const defaultProps: TriStateCheckboxProps = {
    state: 'idle',
    text: 'Test Checkbox',
    onToggle: vi.fn(),
    onMouseDown: vi.fn(),
    id: 'test-checkbox',
    ariaLabel: 'Test Checkbox',
    part: 'checkbox',
    class: 'custom-class',
  };

  const renderComponent = (props: Partial<TriStateCheckboxProps> = {}) => {
    const mergedProps = {...defaultProps, ...props};
    return renderFunctionFixture(
      html`${renderTriStateCheckbox({props: mergedProps})}`
    );
  };

  it('renders the checkbox with default state', async () => {
    await renderComponent();
    const button = page.getByRole('button', {pressed: false});
    await expect.element(button).toBeInTheDocument();
  });

  it('renders the checkbox in "selected" state', async () => {
    await renderComponent({state: 'selected'});
    const button = page.getByRole('button', {pressed: true});
    await expect.element(button).toBeInTheDocument();
  });

  it('renders the checkbox in "excluded" state', async () => {
    await renderComponent({state: 'excluded'});
    // @ts-expect-error: mixed is a valid value for aria-pressed
    const button = page.getByRole('button', {pressed: 'mixed'});
    await expect.element(button).toBeInTheDocument();
  });

  it('calls the onToggle callback when clicked', async () => {
    const onToggleMock = vi.fn();
    await renderComponent({
      onToggle: onToggleMock,
      state: 'idle',
    });
    const button = page.getByRole('button').element();
    button?.dispatchEvent(new MouseEvent('click'));
    expect(onToggleMock).toHaveBeenCalledWith(true);
  });

  it('calls the onMouseDown callback when mouse down event occurs', async () => {
    const onMouseDownMock = vi.fn();
    renderComponent({onMouseDown: onMouseDownMock});

    const button = page.getByRole('button').element();
    button?.dispatchEvent(new MouseEvent('mousedown'));
    expect(onMouseDownMock).toHaveBeenCalled();
  });

  it('applies custom classes and attributes', async () => {
    await renderComponent({
      class: 'custom-class',
      id: 'custom-id',
      ariaLabel: 'Custom Label',
    });
    const button = page.getByRole('button');
    await expect.element(button).toHaveClass('custom-class');
    await expect.element(button).toHaveAttribute('id', 'custom-id');
    await expect.element(button).toHaveAttribute('aria-label', 'Custom Label');
  });

  it('renders the correct icon for "selected" state', async () => {
    const container = await renderComponent({state: 'selected'});
    const icon = container.querySelector('atomic-icon');
    await expect.element(icon).toBeVisible();
    expect(icon?.getAttribute('icon')).toMatchSnapshot();
  });

  it('renders the correct icon for "excluded" state', async () => {
    const container = await renderComponent({state: 'excluded'});
    const icon = container.querySelector('atomic-icon');
    await expect.element(icon).toBeVisible();
    expect(icon?.getAttribute('icon')).toMatchSnapshot();
  });

  it('hides the icon when the state is "idle"', async () => {
    const container = await renderComponent({state: 'idle'});
    const icon = container.querySelector('atomic-icon');
    await expect.element(icon).not.toBeVisible();
  });
});
