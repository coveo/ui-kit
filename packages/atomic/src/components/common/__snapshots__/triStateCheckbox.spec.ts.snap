// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`#renderTriStateCheckbox > renders the correct icon for "excluded" state 1`] = `"<svg viewBox="0 0 22 22"><g transform="matrix(.7071 -.7071 .7071 .7071 -3.142 11)"><path d="m9-3.4h2v26.9h-2z"/><path d="m-3.4 9h26.9v2h-26.9z"/></g></svg>"`;

exports[`#renderTriStateCheckbox > renders the correct icon for "selected" state 1`] = `"<svg viewBox="0 0 12 9" fill="none" xmlns="http://www.w3.org/2000/svg">    <path d="M1.5 5L4.6 7.99999L11 1" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" /></svg>"`;
